File: demo/so-bindings/zig/ffi_example.zig
const c = @cImport({
    @cInclude("example.h");
    @cInclude("stdio.h");
});

fn cstr(s: [*:0]const u8) [*c]u8 {
    return @ptrCast(@constCast(s));
}

pub fn main() void {
    var n: c_longlong = 1;
    while (n <= 10) : (n += 1) {
        _ = c.printf("%lld! = %lld\n", n, c.factorial(n));
    }

    _ = c.printf("%s\n", c.greet(cstr("World")));
    _ = c.printf("%s\n", c.repeat_string(cstr("ha"), 3));

    c.shout_it(cstr("hello from yamlscript"));

    const m = c.maybe();
    _ = c.printf("maybe: %s\n", if (m != 0) cstr("true") else cstr("false"));

    _ = c.printf("sorted: %s\n", c.sort_json_array(cstr("[3,1,4,1,5,9,2,6]")));
}

File: demo/so-bindings/c/ffi_example.c
#include <stdio.h>
#include <stdbool.h>
#include "example.h"

int main() {
    for (long long n = 1; n <= 10; n++)
        printf("%lld! = %lld\n", n, factorial(n));

    printf("%s\n", greet("World"));
    printf("%s\n", repeat_string("ha", 3));

    shout_it("hello from yamlscript");

    printf("maybe: %s\n", maybe() ? "true" : "false");

    printf("sorted: %s\n", sort_json_array("[3,1,4,1,5,9,2,6]"));

    return 0;
}
